### PR TITLE
Smirzaei snow 932731 snowpark sp cancel query

### PR DESCRIPTION
**Pull request by sfc-gh-smirzaei**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

 Using the XP's new end-point for canceling async child jobs from stored proc (based the [python's connector new API for canceling queries](https://github.com/snowflakedb/Stored-Proc-Python-Connector/pull/133)).
NOTE: I'll add regression tests in a separate PR.

Precommit with Snowflake bptp-stable and https://github.com/snowflakedb/Stored-Proc-Python-Connector/pull/133:
https://ci-dev-142.int.snowflakecomputing.com/job/PythonStoredProcBuildPrecommitTest/664/

### CODE REVIEW

File: src/snowflake/snowpark/async_job.py [hunk 1]
```diff
@@ -237,7 +239,19 @@ def is_done(self) -> bool:
...
+                "ENABLE_ASYNC_QUERY_IN_PYTHON_STORED_PROCS", False
+            )
+        ):
+            cancel_resp = self._session._conn._conn.cancel_query(self.query_id)
+            if "success" not in cancel_resp or not cancel_resp["success"]:
+                raise DatabaseError(
+                    f"Failed to cancel query. Returned response: {cancel_resp}"
+                )
+        else:
+            self._cursor.execute(f"select SYSTEM$CANCEL_QUERY('{self.query_id}')")
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/async_job.py:254]: I wonder if we could add a `cancel_query` API in python connector as well. This way we only need to call `cancel_query` from connector?

Comment (sfc-gh-stan) [reply src/snowflake/snowpark/async_job.py:254]: I think we should. There seems to be no public API in the connector to cancel query.

File: src/snowflake/snowpark/async_job.py [hunk 2]
```diff
@@ -237,7 +239,19 @@ def is_done(self) -> bool:
...
         # stop and cancel current query id
-        self._cursor.execute(f"select SYSTEM$CANCEL_QUERY('{self.query_id}')")
+        if (
+            is_in_stored_procedure()
+            and self._session._conn._get_client_side_session_parameter(
+                "ENABLE_ASYNC_QUERY_IN_PYTHON_STORED_PROCS", False
+            )
+        ):
+            cancel_resp = self._session._conn._conn.cancel_query(self.query_id)
+            if "success" not in cancel_resp or not cancel_resp["success"]:
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/async_job.py:249]: NIT: `cancel_resp.get("success", False)`

File: src/snowflake/snowpark/async_job.py (hunk 1 above)
Comment (sfc-gh-smirzaei) [reply src/snowflake/snowpark/async_job.py:254]: Agree we should. I created a separate jira (SNOW-941385) to add the API in public python connector.

**Review by sfc-gh-sfan — APPROVED** [review]

**Pull request merged**